tests/test_url_handler.py

```diff
@@ -0,0 +1,696 @@
+    factories.ProjectMetadataFactory(project=project, metadata=metadata)
+    results = url_handler.get_metadata(project)
+
+    for _, value_iter in results:
+        assert list(value_iter).sort() == vals.sort()
+
+
+@pytest.mark.parametrize('posted_data, processed_posted_data, expected', [
+    (
+        {'color': ['blue', 'other_specify']},
+        {'color': ['blue', 'other_specify'], 'color_other': 'magenta'},
+        {'color': ['blue', 'magenta'], 'color_other': 'magenta'}
+    ),
+    (
+        {'color': ['blue', 'other_specify']},
+        {'color': ['blue', 'other_specify']},
+        {'color': ['blue']}
+    ),
+    (
+        {'color': ['blue', 'green']},
+        {'color': ['blue', 'green']},
+        {'color': ['blue', 'green']}
+    ),
+    (
+        {'color': ['other_specify']},
+        {'color': 'other_specify', 'color_other': 'magenta'},
+        {'color': 'magenta', 'color_other': 'magenta'}
+    ),
+    pytest.mark.xfail((
+        {'color': ['other_specify']},
+        {'color': 'other_specify'},
+        {}
+    ))
+])
+def test_handle_metadata(rf, posted_data, processed_posted_data, expected):
+    request = rf.post('/', posted_data)
+    assert url_handler.handle_metadata(request, processed_posted_data) == expected
+
+
+@pytest.mark.parametrize('date_str', [
+    '2006-10-25',
+    '10/25/2006',
+    '10/25/06',
+    'Oct 25 2006',
+    'Oct 25, 2006',
+    '25 Oct 2006',
+    '25 Oct, 2006',
+    'October 25 2006',
+    'October 25, 2006',
+    '25 October 2006',
+    '25 October, 2006'
+])
+def test_validate_date_with_valid_dates(date_str):
+    assert isinstance(url_handler.validate_date(date_str), datetime.date)
+
+
+def test_validate_date_with_invalid_date():
+    date_str = '2006, Oct 25'
+    assert url_handler.validate_date(date_str) is None
+
+
+def test_add_url():
+    project = factories.ProjectWithMetadataFactory()
+    attribute_names = [x.name for x in project.metadata.all()]
+    value = 'some_value'
+    form_data = {
+        'url_value': 'http://www.example.com',
+        'nominator_email': 'somebody@someplace.com',
+        'nominator_name': 'John Lambda',
+        'nominator_institution': 'UNT',
+        attribute_names[0]: value
+    }
+    expected = [
+        'You have successfully nominated {0}'.format(form_data['url_value']),
+        'You have successfully added the {0} "{1}" for {2}'.format(
+            attribute_names[0], value, form_data['url_value'])
+    ]
+
+    assert url_handler.add_url(project, form_data) == expected
+
+
+def test_add_url_cannot_get_system_nominator():
+    form_data = {'url_value': 'http://www.example.com'}
+    project = factories.ProjectFactory()
+    models.Nominator.objects.get().delete()
+
+    with pytest.raises(http.Http404):
+        url_handler.add_url(project, form_data)
+
+
+@pytest.mark.xfail(reason='Called function surt_exists raises 404, does not return Falsy value.')
+def test_add_url_cannot_get_or_create_surt():
```

Comment:
ok, I'll rename this test once I get to the point of moving everything into classes.